README.md
# azure-graphapi

Node.js package for making Azure Active Directory Graph API calls

v0.0.9

## Installation

```
npm install azure-graphapi --save
```

## Usage

```javascript
var GraphAPI = require('azure-graphapi');

var graph = new GraphAPI(tenant, clientId, clientSecret);
// The tenant, clientId, and clientSecret are usually in a configuration file.

graph.get('users/a8272675-dc21-4ff4-bc8d-8647830fa7db', function(err, user) {
    if (!err) {
        console.dir(user);
    }
}
```

## Details

This package provides an HTTPS interface to the [Azure Active Directory Graph API](https://msdn.microsoft.com/en-us/library/azure/hh974476.aspx). You will need the tenant (i.e., domain) of your Azure AD instance as well as an application within that AD instance that has permissions to access your directory. This application is identified by a `clientId` and authenticated using a `clientSecret`. The `clientSecret` is also called the application key.

The typical verbs are supported (GET, POST, PUT, PATCH, and DELETE). The `getObjects` method is useful for reading a large number of objects. Azure AD limits each response to 100 objects. The `getObject` method follows the `odata.nextLink` and accumulates all objects of a specific object type.

## Interface

This section describes the constructor returned by this module and the available instance methodss.

### Constructor

#### GraphAPI(tenant, clientId, clientSecret, [apiVersion])

Creates a new `GraphAPI` instance. If the `apiVersion` is not specified, it defaults to version 1.5.

```javascript
var GraphAPI = require('azure-graphapi');

var graph = new GraphAPI(tenant, clientId, clientSecret);
```

### Methods

Each of these methods takes `ref` parameter with optional replacement `args`. These are passed to [strformat](https://github.com/fhellwig/strformat) to create the actual URI reference. This allows for the following constructs:

```javascript
var person = {
    userId: 'a8272675-dc21-4ff4-bc8d-8647830fa7db'
};

graph.get('users/{0}', person.userId, function(err, user) {
    if (!err) {
        console.dir(user);
    }
}
```

The following REST-style URI reference is also available:

```javascript
var person = {
    userId: 'a8272675-dc21-4ff4-bc8d-8647830fa7db'
};

graph.get('users/{userId}', person, function(err, user) {
    if (!err) {
        console.dir(user);
    }
}
```

In all cases, do *not* prefix the `ref` with a forward slash and do *not* add the API version as this is added automatically based on the API version specified in the constructor.

#### graph.get(ref [,args...], callback)

Performs an HTTPS GET request. The callback signature is `callback(err, response)`.

#### graph.post(ref [,args...], data, contentType, callback)

Performs an HTTPS POST request. The callback signature is `callback(err, response)`.

#### graph.put(ref [,args...], data, contentType, callback)

Performs an HTTPS PUT request. The callback signature is `callback(err, response)`.

#### graph.patch(ref [,args...], data, contentType, callback)

Performs an HTTPS PATCH request. The callback signature is `callback(err, response)`.

#### graph.delete(ref [,args...], callback)

Performs an HTTPS DELETE request. The callback signature is `callback(err)`.

#### graph.getObjects(ref [,args...], objectType, callback)

Performs an HTTPS GET request and accumulates all objects having the specified `objectType` (e.g., "User"). The callback signature is `callback(err, objects)`. This method follows the `odata.nextLink` property in the response and continues until no more batches of objects are available.

## Notes

1. The HTTPS request logic parses out the `error_description` and `odata.error` messages from JSON responses to unsuccessful requests. These become part of the error message in the `err` object provided to the callback method.
2. If the `contentType` is null and the request data is a string, it is assumed to be form data and is sent as `application/x-www-form-urlencoded`.
3. If the `contentType` is null and the request data is a JavaScript object, it is send as `application/json`.
4. The callback is passed only one argument (err) if the response status code is 204 (No Content). This is important when using libraries such as the [async utilities](https://github.com/caolan/async).
5. If the response has a `value` property, it is that value that is passed to the callback function and not the entire response object.

## License

(The MIT License)

Copyright (c) 2015 Frank Hellwig

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.




CHANGES.md
# azure-graphapi change log

## Version 0.0.9

- Accept a Buffer as request data.

## Version 0.0.8

- Changed strformat dependency from version 0.0.5 to 0.0.7.
- Updated comments.
- Added .DS_Store to .gitignore.

## Version 0.0.7

- Changed the main property in package.json to the correct case.

## Version 0.0.6

- Status code was not attached to error object from httpRequest.

## Version 0.0.5

- Access token was not being set correctly on retry.

## Version 0.0.4

- Fixed bug in the _getObjects method.
- Removed the generic request method.
- Fixed bug when appending version to existing query string.
 
## Version 0.0.3

- Allow string formatting using placeholders (require strformat).
- Fixed bug in getObjects where get was not called with a context.
- Remove log object reference.
- Fixed HTTP status code lookup (require https).

## Version 0.0.2

- Corrected typographical errors in the README.md file.
- Corrected the repository in the package.json file.
- Corrected comments in the GraphAPI.js file.
- The callback data is the response.value and not the response itself.

## Version 0.0.1

- Initial commit.


GraphAPI.js
/**
 * Azure Active Directory (AAD) Graph API
 *
 * Provides an HTTPS interface to the AAD Graph API. This module requests an
 * access token for the application specified in the constructor and then uses
 * that token to make the API calls. If a call fails due to a 401 error, a new
 * new access token is obtained and the request is retried.
 *
 * @author Frank Hellwig
 * @module GraphAPI
 * @version 0.0.8
 */

var http = require('http'),
    https = require('https'),
    querystring = require('querystring'),
    strformat = require('strformat'),
    slice = Array.prototype.slice,
    AAD_LOGIN_HOSTNAME = 'login.windows.net',
    GRAPH_API_HOSTNAME = 'graph.windows.net',
    DEFAULT_API_VERSION = '1.5';

//-----------------------------------------------------------------------------
// PUBLIC
//-----------------------------------------------------------------------------

/**
 * Constructor
 */
function GraphAPI(tenant, clientId, clientSecret, apiVersion) {
    if (!(this instanceof arguments.callee)) {
        throw new Error("Constructor called as a function");
    }
    this.tenant = tenant;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.apiVersion = apiVersion || DEFAULT_API_VERSION;
    this.accessToken = null;
}

/**
 * HTTPS GET
 *
 * get(ref, callback)
 */
GraphAPI.prototype.get = function() {
    var ref = strformat.apply(null, slice.call(arguments, 0, -1)),
        callback = slice.call(arguments, -1)[0];
    this._request('GET', ref, null, null, wrap(callback));
}

/**
 * HTTPS GET with odata.nextList recursive call
 *
 * getObjects(ref, objectType, callback)
 */
GraphAPI.prototype.getObjects = function() {
    var ref = strformat.apply(null, slice.call(arguments, 0, -2)),
        objectType = slice.call(arguments, -2, -1)[0],
        callback = slice.call(arguments, -1)[0];
    this._getObjects(ref, [], objectType, callback);
}

/**
 * HTTPS POST
 *
 * post(ref, data, contentType, callback)
 */
GraphAPI.prototype.post = function() {
    var ref = strformat.apply(null, slice.call(arguments, 0, -3)),
        data = slice.call(arguments, -3, -2)[0],
        contentType = slice.call(arguments, -2, -1)[0],
        callback = slice.call(arguments, -1)[0];
    this._request('POST', ref, data, contentType, wrap(callback));
}

/**
 * HTTPS PUT
 *
 * put(ref, data, contentType, callback)
 */
GraphAPI.prototype.put = function() {
    var ref = strformat.apply(null, slice.call(arguments, 0, -3)),
        data = slice.call(arguments, -3, -2)[0],
        contentType = slice.call(arguments, -2, -1)[0],
        callback = slice.call(arguments, -1)[0];
    this._request('PUT', ref, data, contentType, wrap(callback));
}

/**
 * HTTPS PATCH
 *
 * patch(ref, data, contentType, callback)
 */
GraphAPI.prototype.patch = function() {
    var ref = strformat.apply(null, slice.call(arguments, 0, -3)),
        data = slice.call(arguments, -3, -2)[0],
        contentType = slice.call(arguments, -2, -1)[0],
        callback = slice.call(arguments, -1)[0];
    this._request('PATCH', ref, data, contentType, wrap(callback));
}

/**
 * HTTPS DELETE
 *
 * get(ref, callback)
 */
GraphAPI.prototype.delete = function() {
    var ref = strformat.apply(null, slice.call(arguments, 0, -1)),
        callback = slice.call(arguments, -1)[0];
    this._request('DELETE', ref, null, null, wrap(callback));
}

//-----------------------------------------------------------------------------
// PRIVATE
//-----------------------------------------------------------------------------

// Only return the value and the correct number of arguments.
function wrap(callback) {
    return function(err, response) {
        if (err) {
            callback(err);
        } else if (typeof response === 'undefined') {
            // Handle 204 responses by not adding a second argument.
            callback(null);
        } else if (response.value) {
            callback(null, response.value);
        } else {
            callback(null, response);
        }
    }
}

// Recursive method that follows the odata.nextLink.
GraphAPI.prototype._getObjects = function(ref, objects, objectType, callback) {
    var self = this;
    self._request('GET', ref, null, null, function(err, response) {
        if (err) return callback(err);
        var value = response.value;
        for (var i = 0, n = value.length; i < n; i++) {
            if (value[i].objectType === objectType) {
                objects.push(value[i]);
            }
        }
        var nextLink = response['odata.nextLink'];
        if (nextLink) {
            self._getObjects(nextLink, objects, objectType, callback);
        } else {
            callback(null, objects);
        }
    });
}

// If there is an access token, perform the request. If not, get an
// access token and then perform the request.
GraphAPI.prototype._request = function(method, ref, data, contentType, callback) {
    var self = this;
    if (self.accessToken) {
        self._requestWithRetry(method, ref, data, contentType, false, callback);
    } else {
        self._requestAccessToken(function(err, token) {
            if (err) {
                callback(err);
            } else {
                self.accessToken = token;
                self._requestWithRetry(method, ref, data, contentType, false, callback);
            }
        });
    }
}

// Performs the HTTPS request and tries again on a 401 error
// by getting another access token and repeating the request.
GraphAPI.prototype._requestWithRetry = function(method, ref, data, contentType, secondAttempt, callback) {
    var self = this;
    var path = ['/'];
    path.push(self.tenant);
    path.push('/');
    path.push(ref);
    if (ref.indexOf('?') < 0) {
        path.push('?');
    } else {
        path.push('&');
    }
    path.push('api-version=');
    path.push(self.apiVersion);
    var options = {
        hostname: GRAPH_API_HOSTNAME,
        path: path.join(''),
        method: method,
        headers: {
            'Authorization': 'Bearer ' + self.accessToken
        }
    };
    if (data) {
        if (Buffer.isBuffer(data)) {
            options.headers['Content-Type'] = contentType;
        } else if (!contentType) {
            if (typeof content === 'string') {
                options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
                options.headers['Content-Length'] = content.length;
            } else if (content !== null && typeof content === 'object') {
                content = JSON.stringify(content);
                options.headers['Content-Type'] = 'application/json';
                options.headers['Content-Length'] = content.length;
            }
        }
    }
    httpsRequest(options, data, function(err, response) {
        if (err) {
            if (err.statusCode === 401 && !secondAttempt) {
                self._requestAccessToken(function(err, token) {
                    if (err) {
                        callback(err);
                    } else {
                        self.accessToken = token;
                        self._requestWithRetry(method, ref, data, contentType, true, callback);
                    }
                });
            } else {
                callback(err);
            }
        } else {
            callback(null, response);
        }
    });
}

// Gets an access token using the client id and secret.
GraphAPI.prototype._requestAccessToken = function(callback) {
    var query = {
        client_id: this.clientId,
        client_secret: this.clientSecret,
        grant_type: 'client_credentials',
        resource: 'https://' + GRAPH_API_HOSTNAME
    };
    var content = querystring.stringify(query);
    var options = {
        hostname: AAD_LOGIN_HOSTNAME,
        path: '/' + this.tenant + '/oauth2/token',
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': content.length
        }
    };
    httpsRequest(options, content, function(err, response) {
        if (err) {
            callback(err);
        } else {
            callback(null, response.access_token);
        }
    });
}

// Our own wrapper around the https.request method.
function httpsRequest(options, content, callback) {
    options.headers = options.headers || {};
    options.headers['Accept'] = 'application/json';
    if (!callback) {
        callback = content;
        content = null;
    }
    var req = https.request(options, function(res) {
        res.setEncoding('utf8');
        var buf = [];
        res.on('data', function(data) {
            buf.push(data);
        });
        res.on('end', function() {
            var data = buf.join('');
            if (data.length > 0) {
                data = JSON.parse(data);
            } else {
                data = null;
            }
            if (res.statusCode === 204) {
                callback(null); // no data
            } else if (res.statusCode >= 200 && res.statusCode <= 299) {
                callback(null, data); // success
            } else {
                if (data && data.error_description) {
                    data = data.error_description.split(/[\r\n]/)[0];
                } else if (data && data['odata.error']) {
                    data = data['odata.error'].message.value;
                } else {
                    data = null;
                }
                var err = new Error(errmsg(res.statusCode, data));
                err.statusCode = res.statusCode;
                callback(err);
            }
        });
    });
    req.on('error', function(err) {
        callback(err);
    })
    if (content) {
        req.write(content);
    }
    req.end();
}

// Creates an exception error message.
function errmsg(status, message) {
    message = message || '[no additional details]';
    return strformat('Graph API Error: {0} ({1}) {2}',
        status, http.STATUS_CODES[status], message);
}

//-----------------------------------------------------------------------------
// EXPORTS
//-----------------------------------------------------------------------------

module.exports = GraphAPI; // export the constructor
